keydata.go

```diff
@@ -19,10 +19,506 @@
+// if the keys cannot be successfully recovered because the platform's secure device has
+// not been initialized properly.
+type PlatformUninitializedError struct {
+	err error
+}
+
+func (e *PlatformUninitializedError) Error() string {
+	return fmt.Sprintf("the platform's secure device is not properly initialized: %v", e.err)
+}
+
+func (e *PlatformUninitializedError) Unwrap() error {
+	return e.err
+}
+
+// PlatformDeviceUnavailableError is returned from any of the KeyData.RecoverKeys*
+// functions if the keys cannot be recovered because the platform's secure device is
+// currently unavailable.
+type PlatformDeviceUnavailableError struct {
+	err error
+}
+
+func (e *PlatformDeviceUnavailableError) Error() string {
+	return fmt.Sprintf("the platform's secure device is unavailable: %v", e.err)
+}
+
+func (e *PlatformDeviceUnavailableError) Unwrap() error {
+	return e.err
+}
+
+// DiskUnlockKey is the key used to unlock a LUKS volume.
+type DiskUnlockKey []byte
+
+// AuxiliaryKey is an additional key used to modify properties of a KeyData
+// object without having to create a new object.
+type AuxiliaryKey []byte
+
+// KeyPayload is the payload that should be encrypted by a platform's secure device.
+type KeyPayload []byte
+
+// Unmarshal obtains the keys from this payload.
+func (c KeyPayload) Unmarshal() (key DiskUnlockKey, auxKey AuxiliaryKey, err error) {
+	r := bytes.NewReader(c)
+
+	var sz uint16
+	if err := binary.Read(r, binary.BigEndian, &sz); err != nil {
+		return nil, nil, err
+	}
+
+	if sz > 0 {
+		key = make(DiskUnlockKey, sz)
+		if _, err := r.Read(key); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	if err := binary.Read(r, binary.BigEndian, &sz); err != nil {
+		return nil, nil, err
+	}
+
+	if sz > 0 {
+		auxKey = make(AuxiliaryKey, sz)
+		if _, err := r.Read(auxKey); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	if r.Len() > 0 {
+		return nil, nil, errors.New("excess bytes")
+	}
+
+	return
+}
+
 // AuthMode corresponds to a set of authentication mechanisms.
 type AuthMode uint32
 
 const (
 	AuthModeNone       AuthMode = 0
 	AuthModePassphrase AuthMode = 1 << iota
 )
+
+// KeyCreationData is the data required to create a new KeyData object.
+// It should be produced by a platform implementation.
+type KeyCreationData struct {
+	PlatformKeyData
+	PlatformName string // Name of the platform that produced this data
+
+	// AuxiliaryKey is a key used to authorize changes to the key data.
+	// It must match the key protected inside PlatformKeyData.EncryptedPayload.
+	AuxiliaryKey AuxiliaryKey
+
+	// SnapModelAuthHash is the digest algorithm used for HMACs of Snap
+	// device models, and also the digest algorithm used to produce the
+	// key digest.
+	SnapModelAuthHash crypto.Hash
+}
+
+// KeyID is the unique ID for a KeyData object. It is used to facilitate the
+// sharing of state between the early boot environment and OS runtime.
+type KeyID struct {
```

Comment:
The purpose of KeyID is to provide a mechanism to propagate status from early boot to the runtime, and for the runtime to keep track of keys / revisions so that it can get back to a consistent state if an update fails or is interrupted. It's probably ok to unexport the fields, but I wonder whether I'm overcomplicating this and that it should just be a string instead (or just have a name and a revision). In the case where a key is loaded from LUKS metadata by secboot, the ID would be determined by secboot based on the device node / token name. In the case where a key is loaded externally from a file, the caller (snap-bootstrap in this case) could provide a well known name for the key instead of trying to construct an ID based on a device and path.

Reply:
Ah, I remember why they're exported now - this will eventually go through encoding/json. I've simplified KeyID a bit now (it's just a name/revision pair) and kept the fields exported, but it's passed around as a value now.